doc/specs/#976 - VT52 escape sequences.md

```diff
@@ -0,0 +1,141 @@
+
+Technically the VT52 keyboard doesn't map directly to a typical PC keyboard, so we can't always work from the specs in deciding what sequences are required for each key. When in doubt, we should probably be trying to match the key sequences generated by XTerm.
+
+### Changing Modes
+
+The `_PrivateModeParamsHelper` method in the `AdaptDispatch` class would need to be extended to handle the DECANM mode parameter, and trigger a function to switch to VT52 mode. The typical pattern for this seems to be through a `PrivateXXX` method in the `ConGetSet` interface. Then the `ConhostInternalGetSet` implementation can pass that flag on to the active output buffer's `StateMachine`, and the active input buffer's `TerminalInput` instance.
+
+Changing back from VT52 mode to ANSI mode would need to be achieved with a separate VT52 command, since the VT100 CSI mode sequences would no longer be active. This would be handled in the same place as the other VT52 commands, in the `OutputStateMachineEngine`, and then passed on to the mode selection method in the `AdaptDispatch` class described above (essentially the equivalent of the DECANM private mode being set).
+
+### Additional VT52 Commands
+
+Most of the missing VT52 functionality can be implemented in terms of existing VT100 methods.
+ 
+* The _Cursor Up_, _Cursor Down_, _Cursor Left_, and _Cursor Right_ commands are already implemented.
+* The _Enter Graphics Mode_ and _Exit Graphics Mode_ commands can probably use the existing `DesignateCharset` method, although this would require a new `VTCharacterSets` option with a corresponding table of characters (see below).
+* The _Reverse Line Feed_ command can use the existing `ReverseLineFeed` method.
+* The _Erase to End of Display_ and _Erase to End of Line_ commands can use the existing `EraseInDisplay` and `EraseInLine` methods.
+* The _Cursor Home_ and _Direct Cursor Address_ commands can probably be implemented using the `CursorPosition` method. Technically the _Direct Cursor Address_ has different rules for the boundary conditions, but nobody seems to get that right, so it's probably not that big a deal.
+* The _Identify_ command may be the only one that doesn't build on existing functionality, but it should be a fairly trivial addition to the `AdaptDispatch` class.
+* The _Enter Keypad Mode_ and _Exit Keypad Mode_ commands can use the existing `SetKeypadMode` method, assuming the `Terminal Input` class already knows to generate different sequences when in VT52 mode.
+* The _Enter ANSI Mode_ command can just call through to the new mode selection method in the `AdaptDispatch` class as discussed in the _Changing Modes_ section above.
+
+There are also a few VT52 print commands, but those are not technically part of the core command set, so they can probably be considered out of scope for now.

```

Comment:
The CUP command clamps out of range coordinates, but the _Direct Cursor Address_ command is meant to ignore out of range coordinates, evaluating each individually (so the X may be ignored, while the Y is still applied, and vice versa). I'll add this to the spec.

Reply:
I was thinking they'd likely never be implemented actually. I was somewhat doubtful you'd want to implement any VT52 support, because it doesn't seem likely that there'd be much demand for it. And given that we don't even support the VT100 print commands, I'd think the VT52 ones would be pretty far down the priority list. But I'll add a list of them here.

Reply:
In my dream world, we'd implement literally everything.

The only reason we've skipped things so far is too-much-to-do-and-not-enough-hands. But it is open source now, so I'm not opposed to some ambitious individual implementing things that aren't oft used for the sake of completeness.